src/pyatsimagebuilder/builder.py

```diff
@@ -434,5 +436,28 @@ def _build_image(self, no_cache=False):
         if not self.image.id:
             # we've failed to set the image id - something is wrong!
             raise Exception('No confirmation of successful build.')
-    
 
+    def _replace_environment_variables(self):
+
+        ENV_PATTERN = re.compile(r'(%ENV{ *([0-9a-zA-Z\_]+) *})')
+
+        def _recursive_handle_leaf(data, handle):
+            if isinstance(data, dict):
+                for key in data:
+                    if isinstance(data[key], str):
+                        data[key] = handle(data[key])
+                    elif isinstance(data[key], list):
+                        for i in range(len(data[key])):
+                            data[key][i] = handle(data[key][i])
+                    else:
+                        _recursive_handle_leaf(data[key], handle)
+            else:
+                raise TypeError("Need dict, type={}".format(type(data)))
+
+        def replace_environment_variable(data):
```

Comment:
Try to avoid nesting functions unless you have specific scope requirements. Both `replace_environment_variable` and `_recursive_handle_leaf` can be defined outside of `_replace_environment_variables`.

Reply:
moved them outside of the class